src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -402,6 +402,37 @@ private static bool TryGetClassDeclarationList(INamedTypeSymbol typeSymbol, [Not
                 return typeGenerationSpec;
             }
 
+            internal static bool IsSyntaxTargetForGeneration(SyntaxNode node) => node is ClassDeclarationSyntax { AttributeLists: { Count: > 0 }, BaseList: { Types : {Count : 0 } } };
+
+            internal static ClassDeclarationSyntax? IsSemanticTargetForGeneration(GeneratorSyntaxContext context)
+            {
+                var classDeclarationSyntax = (ClassDeclarationSyntax)context.Node;
+
+                foreach (AttributeListSyntax attributeListSyntax in classDeclarationSyntax.AttributeLists)
+                {
+                    foreach (var attributeSyntax in attributeListSyntax.Attributes)
+                    {
+                        IMethodSymbol attributeSymbol = context.SemanticModel.GetSymbolInfo(attributeSyntax).Symbol as IMethodSymbol;
+                        if (attributeSymbol == null)
+                        {
+                            continue;
+                        }
+
+                        INamedTypeSymbol attributeContainingTypeSymbol = attributeSymbol.ContainingType;
+                        var fullName = attributeContainingTypeSymbol.ToDisplayString();
+
+                        if (fullName == "System.Text.Json.Serialization.JsonSerializableAttribute"
+                            || fullName == "System.Text.Json.Serialization.JsonSourceGenerationOptionsAttribute")
```

Comment:
I _think_ we just need to check for `JsonSerializableAttribute`. Having `JsonSourceGenerationOptionsAttribute` by itself on a class doesn't do anything, AFAIK.